pkg/common/cns-lib/volume/manager.go

```diff
@@ -515,6 +564,79 @@ func (m *defaultManager) createVolumeWithImprovedIdempotency(ctx context.Context
 	return resp, faultType, err
 }
 
+func (m *defaultManager) waitOnTask(csiOpContext context.Context,
+	taskMoRef vim25types.ManagedObjectReference) (*vim25types.TaskInfo, error) {
+	log := logger.GetLogger(csiOpContext)
+	if m.listViewIf == nil {
+		err := m.initListView()
+		if err != nil {
+			return nil, err
+		}
+	}
+	ch := make(chan TaskResult)
+	err := m.listViewIf.AddTask(csiOpContext, taskMoRef, ch)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "failed to add task to list view. err: %v", err)
+	}
+
+	// deferring removal of task after response from CNS
+	// called only after successful addition to listView, so we don't need any check for the removal
+	defer func() {
+		err := m.listViewIf.RemoveTask(csiOpContext, taskMoRef)
+		if err != nil {
+			log.Errorf("failed to remove task from list view. err: %v", err)
+			err = m.listViewIf.MarkTaskForDeletion(csiOpContext, taskMoRef)
+			if err != nil {
+				log.Errorf("failed to mark task for deletion. err: %v", err)
+			}
+		}
+	}()
+
+	return waitForResultOrTimeout(csiOpContext, ch)
+}
+
+// waitForResultOrTimeout uses the context provided by the sidecars when CSI driver operations are called.
+// This context has a timeout associated with it (see manifests for more details).
+// Once this caller timeout is over, we want to return an error back to the caller
+func waitForResultOrTimeout(csiOpContext context.Context, ch chan TaskResult) (*vim25types.TaskInfo, error) {
```

Comment:
You can make this method a bit more informative. Pass in the taskmoref from waitOnTask, and in case there was an error, your error string can explicitly mention which task timedout. This will make triage easier.

Reply:
Thanks! I've added the task ref to the error being returned.